detox/src/utils/ExclusiveLockfile.js

```diff
@@ -32,8 +32,11 @@ class ExclusiveLockfile {
    * @returns {Promise<any>}
    */
   async exclusively(fn) {
-    await this._lock();
+    if (this._isLocked) {
+      return await fn();
```

Comment:
⚠️  No, wait, what are you doing?? If it is locked, you should stop at line `await this._lock()` ⚠️ That's the idea.

@d4vidi ??

Reply:
Wow thanks for going over this code so soon 😅 ❤️  
I think the terms here are misleading - the logic is a bit reversed: `_isLocked=true` means you've already "obtained the semaphore" and that you can do whatever you like. Only otherwise, you are forced to go into `this._lock()` and wait for your turn, inside.

Reply:
Your instinct is correct, but do note that there's a unit test that validates this behavior (i.e. can't execute actions if not locked): https://github.com/wix/Detox/blob/611ae1aaa1f964407a10237554197c7fd95da6d8/detox/src/utils/ExclusiveLockfile.test.js#L50

Reply:
@d4vidi, no, there's something really odd here. First of all, why did you have to change that place at all? What happened that you had to change something in that class?

My initial idea was that there can be *no condition* in our source code, when two calls of `.exclusively(callback)` are being executed. Every call when it ends the execution, unlocks the file, so the next one can start.

Now, with your change, we can start having race conditions inside our source code, with the lock file. I am not convinced that this change of yours is safe. Could you elaborate more, please?

Reply:
Oh yes I see now - you don't want potential simultaneous callers to step on each other's toes, eventually polluting the value of the shared resource.

Well, I have another solution, which isn't as nice, but will do nonetheless. Let's discuss 1x1.

Reply:
@d4vidi, well, I did not want it to happen just **accidentally**. Let's discuss it indeed.